units/generate_tasks.go

```diff
@@ -93,8 +94,8 @@ func (j *generateTasksJob) generate(ctx context.Context, t *task.Task) error {
 	// SIGTERM from a deploy. This should maybe be a context with timeout.
 	err = g.Save(context.Background(), p, pp, v, t, pm)
 
-	// If the version has changed there was a race. Another generator will try again.
-	if adb.ResultsNotFound(err) {
+	// If the version or parser project has changed there was a race. Another generator will try again.
+	if adb.ResultsNotFound(err) || db.IsDuplicateKey(err) {
```

Comment:
Where would the duplicate key error come from?

Reply:
This is explained also in the pull request description: a duplicate key error is equivalent to resultsNotFound for parser project, because we're doing an upsert: we attempt to update a parser project matching `id` and `configNum`. If we can't find this parser project, we attempt to insert a new one--this is correct behavior in the case that no parser project for this `id` exists yet, and we should insert a new document. However, if the project in the DB has a different configNum (i.e. a race), we'll still try to insert a new document bc that's how Upsert operates, and this will return a duplicate key error, because the `id` already exists.